[mlir][wasm] Add draft example of support for trappable operation

flemairen6 opened the pull request:
Draft idea on how to support trappable ops:
- Add an interface to mark trappable operations
- Remove the trappable operations from RaiseWasm
- Create a new pass WasmToEmbedder (Name can be discussed - format of API too) - the pass creates forward declaration of calls that would replace the trappable operation and replace the op itself by a call to the forward declared call.

Commit:
[mlir][wasm] Add draft example of support for trappable operation

Commit:
[mlir][wasm] Add specific check in interface to separate trappable from non-trappable ops

lforg37 commented on mlir/include/mlir/Dialect/WebAssembly/IR/WebAssemblyInterfaces.td:
```diff
@@ -183,4 +183,11 @@ def WasmConstantExprInterface :
+// Trait to mark operators that can trigger a trap. Examples: Integer overflow, division
+// by zero, failed conversion...
+def WasmTrappableOpInterface :
+          OpInterface<"WasmTrappableOpInterface"> {}
```
WasmEmbedderDependantOpInterface?

lforg37 commented on mlir/include/mlir/Dialect/WebAssembly/IR/WebAssemblyOps.td:
add is not modified: is that expected?

lforg37 commented on mlir/test/Conversion/WasmMLIRToEmbedder/wasm-add-embedder-add.mlir:
```diff
@@ -1,80 +1,90 @@
-%v1 = wasm.local_get %arg1 : ref to i32
-// CHECK:           %[[VAL_4:.*]] = arith.subi %[[VAL_2]], %[[VAL_3]] : i32
-%res = wasm.sub %v0 %v1 : i32
+// CHECK:           %[[VAL_4:.*]] = call @wasm.add_i32(%[[VAL_2]], %[[VAL_3]]) : (i32, i32) -> i32
```
shouldn't we declare this symbol somewhere?

lforg37 replied:
Ok it is actually defined just above, sorry for the noise

lforg37 reviewed (commented):
A small remark on naming: I think we want to make clear that this WasmMLIRToEmbedder is not a "universal" pass but should be duplicated for each embedder we want to support.

Commit:
[mlir][wasm] Dummy conversion for trapping arithmetic ops

Check overflow result for arithmetic ops that can overflow and trap if
an overflow occured.

Same for integer division if divider is zero.

Commit:
[mlir][wasm] Introduce trap Op and move the lowering logic of trappable ops back to RaiseWasm.

flemairen6 commented:
Replaced by https://github.com/lforg37/llvm-project/pull/29
